common/ApplicationConfig.cc.in

```diff
@@ -17,162 +17,611 @@
 
 const std::string Config::CMSTkModLabBasePath("@basepath@");
 
+bool operator<(const ApplicationConfig::FullKey& k1, const ApplicationConfig::FullKey& k2)
+{
+  if (k1.alias == k2.alias) {
+    return k1.key < k2.key;
+  } else {
+    return k1.alias < k2.alias;
+  }
+}
+
+bool operator<(const ApplicationConfig::FullKey& fk, const ApplicationConfig::ShortKey& sk)
+{
+  if (fk.alias == sk.alias) {
+    return fk.key < sk.key;
+  }
+  return fk.alias < sk.alias;
+}
+
+bool operator<(const ApplicationConfig::ShortKey& sk, const ApplicationConfig::FullKey& fk)
+{
+  if (sk.alias == fk.alias) {
+    return sk.key < fk.key;
+  }
+  return sk.alias < fk.alias;
+}
+
 ApplicationConfig* ApplicationConfig::instance_ = NULL;
 
-ApplicationConfig::ApplicationConfig(const std::string filename)
+ApplicationConfig::ApplicationConfig(QObject *parent)
+ : QObject(parent)
 {
-  ApplicationConfigReader reader(filename);
-  reader.fill(keyvalueMap_);
+
+}
+
+ApplicationConfig* ApplicationConfig::instance(QObject *parent)
+{
+  if (instance_==NULL) {
+    instance_ = new ApplicationConfig(parent);
+  }
+  return instance_;
 }
 
-ApplicationConfig* ApplicationConfig::instance(const std::string filename)
+ApplicationConfig* ApplicationConfig::instance(const std::string filename,
+  const std::string alias, QObject *parent)
 {
   if (instance_==NULL) {
+    instance_ = new ApplicationConfig(parent);
     if (!filename.empty()) {
-      instance_ = new ApplicationConfig(filename);
+      instance_->append(filename, alias);
     }
   }
   return instance_;
 }
 
-void ApplicationConfig::safe(const std::string filename)
+void ApplicationConfig::append(const std::string& filename, const std::string& alias)
+{
+  ApplicationConfigReader reader(filename, alias);
+  reader.fill(keyvalueMap_);
+  
+  emit structureChanged();
+  emit valueChanged();
+}
+
+void ApplicationConfig::save()
+{
+  std::map<std::string,std::string> filemap;
+  for (const auto & kv : keyvalueMap_) {
+  	if (kv.first.alias=="*") continue;
+  	filemap[kv.first.filename] = kv.first.alias;
+  }
+
+  for (const auto & kv : filemap) {
+  	ApplicationConfigWriter writer(kv.first, kv.second);
+    writer.write(keyvalueMap_);
+  }
+}
+
+void ApplicationConfig::saveAs(const std::string& filename, const std::string& alias)
 {
-  ApplicationConfigWriter writer(filename);
-  writer.write(keyvalueMap_);
+  storage_t tempKeyValueMap;
+  
+  FullKey fk;
+  fk.filename = filename;
+  fk.alias = alias;
+  
+  for (const auto & kv : keyvalueMap_) {
+  	if (kv.first.alias!="*") continue;
```

Comment:
This selects all `*` parameters, which are all parameters, as in every parameter is duplicated with a `*` alias. I would expect `saveAs(filename, alias)` to write only the parameters for a certain alias to be written to this file. I already fixed it - if you agree I will change it within a PR to your repository together with all my changes for the assembly code.

Reply:
My thinking was that saveAs writes everything (ignoring aliases) to a new file and assigns a new alias. But I agree that the logic proposed by you makes more sense.

Reply:
I understand. Is there a need for this functionality (overwriting/changing the alias/filename) for you?

Reply:
Fixed in latest commits.